app/pki/pki.go

```diff
@@ -0,0 +1,378 @@
+	//
+	if err := p.verifyCertPEM(p.selfCA.certPEM); err != nil {
+		return fmt.Errorf("p.verifyCertPEM selfCA: %w", err)
+	}
+
+	if err := p.verifyCertPEM(p.ed25519.certPEM); err != nil {
+		return fmt.Errorf("p.verifyCertPEM ed25519: %w", err)
+	}
+
+	if err := p.verifyCertPEM(p.ecdsa.certPEM); err != nil {
+		return fmt.Errorf("p.verifyCertPEM ecdsa: %w", err)
+	}
+
+	if err := p.verifyCertPEM(p.rsa.certPEM); err != nil {
+		return fmt.Errorf("p.verifyCertPEM rsa: %w", err)
+	}
+
+	return nil
+}
+
+func (p *PKI) verifyCertPEM(certPEM []byte) error {
+	// https://pkg.go.dev/crypto/x509#example-Certificate.Verify (4/4)
+	//
+	block, _ := pem.Decode(certPEM)
+	if block == nil {
+		return fmt.Errorf("pem.Decode: block is nil") //nolint:goerr113
+	}
+
+	cert, err := x509.ParseCertificate(block.Bytes)
+	if err != nil {
+		return fmt.Errorf("x509.ParseCertificate: %w", err)
+	}
+
+	opts := x509.VerifyOptions{ //nolint:exhaustruct
+		// DNSName: "",
+		Roots: p.roots,
+	}
+
+	if _, err := cert.Verify(opts); err != nil {
+		return fmt.Errorf("cert.Verify: %w", err)
+	}
+
+	return nil
+}
+
+func (p *PKI) ValidateTLSX509KeyPairCert() error {
+	if err := p.selfCA.tlsX509KeyPairCert(); err != nil {
+		return fmt.Errorf("selfCA.tlsX509KeyPairCert: %w", err)
+	}
+
+	if err := p.ed25519.tlsX509KeyPairCert(); err != nil {
+		return fmt.Errorf("ed25519.tlsX509KeyPairCert: %w", err)
+	}
+
+	if err := p.ecdsa.tlsX509KeyPairCert(); err != nil {
+		return fmt.Errorf("ecdsa.tlsX509KeyPairCert: %w", err)
+	}
+
+	if err := p.rsa.tlsX509KeyPairCert(); err != nil {
+		return fmt.Errorf("rsa.tlsX509KeyPairCert: %w", err)
+	}
+
+	return nil
+}
+
+func (p *PKI) saveKeyPairCerts() error {
+	// switch casePKI {
+	// case _selfCA:
+	// case _ed25519:
+	// case _ecdsa:
+	// case _rsa:
+	// default:
+	// 	return fmt.Errorf("pki: casePKI: %v", casePKI) //nolint:goerr113
+	// }
+	if err := p.saveKeyPairCert(_selfCA, p.selfCA.certPEM, p.selfCA.keyPEM); err != nil {
+		return fmt.Errorf("pki: p.saveKeyPairCert: selfCA %w", err)
+	}
+
+	if err := p.saveKeyPairCert(_ed25519, p.ed25519.certPEM, p.ed25519.keyPEM); err != nil {
+		return fmt.Errorf("pki: p.saveKeyPairCert: ed25519 %w", err)
+	}
+
+	if err := p.saveKeyPairCert(_ecdsa, p.ecdsa.certPEM, p.ecdsa.keyPEM); err != nil {
+		return fmt.Errorf("pki: p.saveKeyPairCert: ecdsa %w", err)
+	}
+
+	if err := p.saveKeyPairCert(_rsa, p.rsa.certPEM, p.rsa.keyPEM); err != nil {
+		return fmt.Errorf("pki: p.saveKeyPairCert: rsa %w", err)
+	}
+
+	return nil
+}
+
+func (p *PKI) saveKeyPairCert(casePKI casePKI, certPEMBlock []byte, keyPEMBlock []byte) error {
+	const (
+		permDir fs.FileMode = 0740 // 740 rwxr-----
+		permCer fs.FileMode = 0640 // 640 rw-r-----
+		permKey fs.FileMode = 0600 // 600 rw-------
+	)
+
+	saveDir := filepath.Join("test", "certs", "auto-self")
+	saveCer := filepath.Join(saveDir, fmt.Sprintf("%s-cert.pem", casePKI))
+	saveKey := filepath.Join(saveDir, fmt.Sprintf("%s-key.pem", casePKI))
```

Comment:
Do you think it is a good idea to use files with microservices? How to do this correctly?

Reply:
It is not the best idea and app/pki doesn't use these files as source but each time overwrite app/pki/test/certs/auto-self. They are to provide content for potential tests of particular case/approach.

How to do this correctly? - It is configured already to use for test purposes on unknown environment with tls switched on with highest crypto standard as possible. Other particular usage depends on the environment and on what we want to achieve. There is no one answer good for all cases.

app/pki simply generates 3 types of key and self sign them.
rest gw and internal cli.grpc use pki.TLSConfigDial() with self-signed cert added to the CertPool() and InsecureSkipVerify is default false so transport is encrypted and server is verified; 
So far they don't use own client's key and server is configured as tls.VerifyClientCertIfGiven - it means server don't verify client;
External grpc cli (like one in fake_test) run on separate os process/instance so it doesn't have the same self-signed certPool and tls config is set to not verify server cert client.InsecureClientConnGRPC() with InsecureSkipVerify: true

It can be enhanced to certify and verify rest gw and client too; 
It can be further modified to certify with external provider (let's encrypt would require external ip) so grpc client in fake could be switched to InsecureSkipVerify: false but it would also require a mechanism of sharing cert or mechanism of certification using cert key corresponding to externally certified pub (last one is recommended)

Entire app is more a study and entry point for discussion or/and tests based on working tool then selected case and path. Also TLS changes dynamically but not all tools uses latest achievements so pki can be used to test periodically where which standard can be switched on.